### DISCUSSION TITLE
Support for Faust

### DISCUSSION THREAD

**Discussion by pschatzmann** [pschatzmann/arduino-audio-tools#171] [Announcements]
I have added some DRAFT support for [Faust](https://faust.grame.fr/). Further information can be found in [my blog](https://www.pschatzmann.ch/home/2022/04/22/using-faust-dsp-with-my-arduino-audio-tools/).

**Comment by jonrichings** [comment]
Hi Phil,

Taking a break from the touch sensor problem I have been trying out the Faust example streams-faust-i2c.  Works just.  Compiles without error and plays a nice soft white noise.  

So then I compiled the default .dsp in the online Faust IDE. and copied those files to the example folder.  Then changed the _#include "Noise.h"_ to _#include "myFirstFaust.h"_.  I didn't think it would be that easy and so Arduino compiler complained about the name.  I tried replacing _mydsp dsp;_ with _myFirstFaust dsp;_ but still no luck.  I suspect this is just a matter of getting the names right.  Any advice?

**Comment by pschatzmann** [comment]
Did you select 
- Platform: source and 
- Architecture: cplusplus ?

This should generate just one file. If you check the generated cpp file it should just have one class which is a subclass of dsp.
ps. I did my examples with Faust Live

**Reply by jonrichings** [reply]
In Faust IDE how do you get the .h that the .ino #includes or is .cpp just as good?
In FaustLive (I hadn't tried that till now), what Compiler Options do you use?

**Comment by pschatzmann** [comment]
I just renamed the .cpp file. Including cpp files is discouraged.
With my latest version you don't really need any special compiler options.

**Comment by jonrichings** [comment]
OK, I'll change them to .h
I'm trying to use FaustLive to compile.  It's not compiling at the moment (maybe their server is down).  
I'm using Platform: Source and Architecture: cplusplus on FaustLive and and FaustIDE.  That should be OK right?
I tried other DSP examples but some of them did produce more than one class.

**Comment by jonrichings** [comment]
I'm getting some of the Faust examples to compile and run.  I re-compiled noise.dsp in FaustIDE and that works so I have at least one example of the complete end-to-end process proven out.

I've been trying various other examples, some fail to compile in the FaustIDE, most work in FaustIDE but fail when added to your .ino and compiled in Arduino.  One or two have compiled and downloaded but crash the ESP32.  Unless you have any specific requests for me I'll just keep playing with different examples and make notes on the results for you.

BTW, I've added the BLE MIDI code into your Faust example.  How can I send note info to Faust's "sliders"?  I tried a simple fHslider0=freq but, not surprisingly got a "'fHslider0' was not declared in this scope" error.

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-generator-faust-i2s/streams-generator-faust-i2s.ino shows how to set parameter values. You will need to check in your generated code how the label is called...

**Reply by jonrichings** [reply]
Sorry Phil, I should have looked at that example.  Anything I can do to help test the Faust integration?

**Comment by jonrichings** [comment]
streams-generator-fause-i2s compiles and generates the expected sine wave.  But, if I'm understanding the purpose, faust.setLabelValue("0x00", 0.5); isn't changing the volume (or anything as far as I can tell).  No audible difference with faust.setLabelValue("0x00", 0.1);

**Comment by pschatzmann** [comment]
If you check the defintion in the source code:
ui_interface->addVerticalSlider("0x00", &fVslider0, FAUSTFLOAT(0.0f), FAUSTFLOAT(-70.0f), FAUSTFLOAT(4.0f), FAUSTFLOAT(0.100000001f));
The range is from -70.0 to 4.0
I suggest to try with a bigger change then just 0.1...

**Reply by jonrichings** [reply]
Actually I tried various values from -70 and 4 and just retested with -70.0, -50.0, -0.5, -5, 0.5, 4.0.  All produce the same volume.

**Comment by pschatzmann** [comment]
Oh, I think the sequence is wrong:   faust.setLabelValue("0x00", 0.5); should be called after begin() because begin() will reset the value to the initial value, which has been defined in addVerticalSlider()

**Reply by jonrichings** [reply]
Thats it!  Thanks Phil :)
